### PR TITLE
Remove Cells or Faces

### PR DESCRIPTION
**Pull request by akaszynski**
### Remove Cells or Faces
Someone posted an example on pyvista-support ([Issue 144](https://github.com/pyvista/pyvista-support/issues/144)), and I found it really inconvenient that we don't have a way of removing cells or faces in-place.

Turns out using ghost cells is the way to go.  See the issue above for a good demo on how it works in practice.

### CODE REVIEW

**Comment by akaszynski** [comment]
@pyvista/developers, ready for review.  Once this is done I'll add in another example using this on the docs.

File: pyvista/core/filters.py [hunk 1]
```diff
@@ -1747,7 +1747,6 @@ def plot_over_line(dataset, pointa, pointb, resolution=None, scalars=None,
         if show:
             return plt.show()
 
-
```
Comment (imsodin) [inline pyvista/core/filters.py]: Just a personal note: I'd put such unrelated formatting into a separate commit and just either push that directly to master, as it's obviously harmless, or create a separate PR. It just adds pointless noise to the diff making a review just a bit less comfortable.

File: pyvista/core/pointset.py [hunk 2]
```diff
@@ -48,14 +48,45 @@ def center_of_mass(self, scalars_weight=False):
...
 
-
     def shallow_copy(self, to_copy):
         """Do a shallow copy the pointset."""
         # Set default points if needed
         if not to_copy.GetPoints():
             to_copy.SetPoints(vtk.vtkPoints())
         return Common.shallow_copy(self, to_copy)
 
+    def remove_cells(self, ind):
```
Comment (imsodin) [inline pyvista/core/pointset.py]: Shouldn't this be a filter? I get that it's not a vtk filter, but semantically it feels like a filter to me (give in a mesh, get out a filtered mesh).

File: pyvista/core/pointset.py [hunk 3]
```diff
@@ -48,14 +48,45 @@ def center_of_mass(self, scalars_weight=False):
...
+        """Remove cells.
+
+        This modifies the mesh or grid in-place without making a copy.
+
+        Parameters
+        ----------
+        ind : np.ndarray, list
```
Comment (imsodin) [inline pyvista/core/pointset.py]: Maybe just `iterable` as type.

File: pyvista/core/pointset.py [hunk 4]
```diff
@@ -478,6 +503,28 @@ def n_open_edges(self):
         alg.Update()
         return alg.GetOutput().GetNumberOfCells()
 
+    def remove_faces(self, ind):
```
Comment (imsodin) [inline pyvista/core/pointset.py]: I'd remove this method in favor of Zen's

> There should be one-- and preferably only one --obvious way to do it.

I think `remove_cells` is understandable in the context of a surface mesh as well.

File: pyvista/core/pointset.py [hunk 5]
```diff
@@ -743,6 +786,11 @@ def save(self, filename, binary=True):
         one system may not be readable on other systems.  Binary can be used
         only ".vtk" files
 
+        Examples
+        --------
+        >>> import pyvista
+        >>> beam = pyvista.UnstructuredGrid(pyvista.examples.hexbeamfile)
+        >>> _ = beam.save('/tmp/tmp.vtk')  # or "C:/Users/where/tmp.vtk"
```
Comment (imsodin) [inline pyvista/core/pointset.py]: This definitely should go into it's own commit. If there's something wrong with the rest and someone feels the need to revert, this should not get reverted.

File: pyvista/core/pointset.py (hunk 4 above)
Comment (akaszynski) [reply pyvista/core/pointset.py]: Agreed.  This also means that we should remove `faces`.  This will be a significant refactor.

File: pyvista/core/pointset.py (hunk 5 above)
Comment (akaszynski) [reply pyvista/core/pointset.py]: While I agree that this is outside the scope of this PR, we're constraining ourselves to lots of PRs (and 30 minutes of build time plus reviews).  I'll remove it here and make another PR in the future.

File: pyvista/core/pointset.py (hunk 3 above)
Comment (akaszynski) [reply pyvista/core/pointset.py]: Agreed.

File: pyvista/core/filters.py (hunk 1 above)
Comment (akaszynski) [reply pyvista/core/filters.py]: True enough.  I'll avoid doing that in the future.

File: pyvista/core/pointset.py (hunk 2 above)
Comment (akaszynski) [reply pyvista/core/pointset.py]: It's not quite a filter, but we can change it such that it acts like a filter with an `in-place` method.  Still, I think it should remain a method in `PointSet`.

I'm not actually a fan of `filters.py`, and would prefer that these methods be added directly to the correct class instead.  The filters file is 3000+ LOC, and if we broke up the modules and had one or two classes per module, I think things would be a bit cleaner as we're importing several subclasses for each main class.

But that's a discussion for a later time.

Comment (imsodin) [reply pyvista/core/pointset.py]: I totally agree on filters.py being too large. But for the same reason I like the distinction of methods on classes directly, which deals with loading/saving, accessing data and other properties, and methods in filters, which modify the objects.

Comment (akaszynski) [reply pyvista/core/pointset.py]: I'm actually thinking that something like this would be better:
http://www.qtrac.eu/pyclassmulti.html

```python
# _DataStore.py

def bigMethod(state):
    ... # lots of lines

def hugeMethod(state):
    ... # lots of lines
```

```python
# DataStore.py
import _DataStore

class DataStore:

    ... # lots of other methods ...

    def bigMethod(self):
	_DataStore.bigMethod(self)

    def hugeMethod(self):
	_DataStore.hugeMethod(self)
```

In fact, we could even have it so that some of these methods could be called directly from `pyvista`.  This would be more in-line with what `numpy` does.  For examples:
```python
import numpy as np

arr = np.arange(10)
print(arr.max())
print(np.max(arr))
```

We'd have to implement type checking within our functions, but I think that's worth it to be able to call the function directly from `pyvista`.  From there we could run `pyvista.remove_cells`.

If not, I'd rather place methods all within the class or common classes, or place the methods in a separate module and then call them as in above.  This would make for some clean code, especially with `@wraps`.

Comment (akaszynski) [reply pyvista/core/pointset.py]: Additionally, I can't really place `remove_cells` in `filters.py` as there isn't a class that  that has methods for just `PolyData` and `UnstructuredGrid`.  Rather, there's a class for all types, `PolyData`, or `UnstructuredGrid`.

I really would prefer to refactor this, and since @banesullivan implemented this, I'll wait until he comes back online to talk about this.  For the time being, I'm leaving this in `PointSet`

**Comment by akaszynski** [comment]
@imsodin, please let me know if you have any more comments.

File: pyvista/core/pointset.py [hunk 6]
```diff
@@ -1054,3 +1093,39 @@ def _get_attrs(self):
         attrs = PointGrid._get_attrs(self)
         attrs.append(("Dimensions", self.dimensions, "{:d}, {:d}, {:d}"))
         return attrs
+
+    def hide_cells(self, ind):
```
Comment (imsodin) [inline pyvista/core/pointset.py]: Why is this on `StructuredGrid` and the removal method on `PointSet`? It might even make sense to merge the two adding a `hide=False` parameter.

Comment (akaszynski) [reply pyvista/core/pointset.py]: I discovered through trial and error that you can't hide cells in an `UnstrucutredGrid` or `PolyData`, but you can in a `StructuredGrid`.  I'd prefer having the same functionality with both the classes, but that's not the way `vtk` is built.

Comment (imsodin) [reply pyvista/core/pointset.py]: That definitely deserves a comment - such obscure information tends to come back to bite you and then such a comment may save quite a bit of time.

Comment (akaszynski) [reply pyvista/core/pointset.py]: Added within the method.

File: tests/test_grid.py [hunk 7]
```diff
@@ -472,3 +472,31 @@ def test_gaussian_smooth():
...
+def test_remove_cells_invalid():
+    grid_copy = beam.copy()
+    with pytest.raises(ValueError):
+        grid_copy.remove_cells(np.ones(10, np.bool))
+
+
+def test_hide_cells():
+    sgrid_copy = sgrid.copy()
+    sgrid_copy.hide_cells(range(10, 20))
+    sgrid_copy.UpdateCellGhostArrayCache()
```
Comment (imsodin) [inline tests/test_grid.py]: Why is this necessary, respectively shouldn't this happen in `hide_cells`?

File: pyvista/core/pointset.py [hunk 8]
```diff
@@ -1054,3 +1094,45 @@ def _get_attrs(self):
...
+        >>> grid.hide_cells(range(79*30, 79*50))
+        """
+        if isinstance(ind, np.ndarray) and ind.dtype == np.bool:
+            if ind.size != self.n_cells:
+                raise ValueError('Boolean array size must match the '
+                                 'number of cells (%d)' % self.n_cells)
+            ghost_cells = ind.astype(np.uint8)
+        else:
+            ghost_cells = np.zeros(self.n_cells, np.uint8)
+            ghost_cells[ind] = vtk.vtkDataSetAttributes.HIDDENCELL
```
Comment (imsodin) [inline pyvista/core/pointset.py]:
```suggestion
        ghost_cells = np.zeros(self.n_cells, np.uint8)
        ghost_cells[ind] = vtk.vtkDataSetAttributes.HIDDENCELL
```

File: pyvista/core/pointset.py [hunk 9]
```diff
@@ -56,6 +56,46 @@ def shallow_copy(self, to_copy):
...
+        >>> letter_a.remove_cells(range(1000))
+        """
+        if isinstance(ind, np.ndarray) and ind.dtype == np.bool:
+            if ind.size != self.n_cells:
+                raise ValueError('Boolean array size must match the '
+                                 'number of cells (%d)' % self.n_cells)
+            ghost_cells = ind.astype(np.uint8)
+        else:
+            ghost_cells = np.zeros(self.n_cells, np.uint8)
+            ghost_cells[ind] = vtk.vtkDataSetAttributes.DUPLICATECELL
```
Comment (imsodin) [inline pyvista/core/pointset.py]:
```suggestion
        ghost_cells = np.zeros(self.n_cells, np.uint8)
        ghost_cells[ind] = vtk.vtkDataSetAttributes.DUPLICATECELL
```

File: pyvista/core/pointset.py [hunk 10]
```diff
@@ -1054,3 +1094,45 @@ def _get_attrs(self):
...
+    def hide_cells(self, ind):
+        """Hide cells without deleting them.
+
+        Hides cells by setting the ghost_cells array to HIDDEN_CELL.
+
+        Parameters
+        ----------
+        ind : iterable
+            List or array of cell indices to be hidden.  The array can
+            also be a boolean array the same size as the number of
```
Comment (imsodin) [inline pyvista/core/pointset.py]:
```suggestion
            also be a boolean array of the same size as the number of
```

File: pyvista/core/pointset.py [hunk 11]
```diff
@@ -1054,3 +1094,45 @@ def _get_attrs(self):
...
+            cells.
+
+        Examples
+        --------
+        Hide part of the middle of a structured surface
```
Comment (imsodin) [inline pyvista/core/pointset.py]:
```suggestion
        Hide part of the middle of a structured surface.
```

File: pyvista/core/pointset.py (hunk 9 above)
Comment (akaszynski) [reply pyvista/core/pointset.py]: Good change.  Thanks!

File: tests/test_grid.py (hunk 7 above)
Comment (akaszynski) [reply tests/test_grid.py]: The array cache is automatically updated when setting `hide_cells`.  Thanks for catching that.

File: pyvista/core/pointset.py [hunk 12]
```diff
@@ -56,6 +56,44 @@ def shallow_copy(self, to_copy):
...
+        Examples
+        --------
+        Remove first 1000 cells from an unstructured grid.
+
+        >>> import pyvista
+        >>> letter_a = pyvista.examples.download_letter_a()
+        >>> letter_a.remove_cells(range(1000))
+        """
+        if isinstance(ind, np.ndarray) and ind.dtype == np.bool:
+            if ind.size != self.n_cells:
```
Comment (imsodin) [inline pyvista/core/pointset.py]: This could now be a single if condition.

File: pyvista/core/pointset.py [hunk 13]
```diff
@@ -1054,3 +1092,43 @@ def _get_attrs(self):
...
+        >>> import numpy as np
+        >>> x = np.arange(-10, 10, 0.25)
+        >>> y = np.arange(-10, 10, 0.25)
+        >>> z = 0
+        >>> x, y, z = np.meshgrid(x, y, z)
+        >>> grid = pv.StructuredGrid(x, y, z)
+        >>> grid.hide_cells(range(79*30, 79*50))
+        """
+        if isinstance(ind, np.ndarray) and ind.dtype == np.bool:
+            if ind.size != self.n_cells:
```
Comment (imsodin) [inline pyvista/core/pointset.py]: This could now be a single if condition.

File: pyvista/core/pointset.py (hunk 12 above)
Comment (akaszynski) [reply pyvista/core/pointset.py]: Maybe I don't see it, but I think that it should remain a nested condition.  Having three conditions on one line is a bit messy.  In fact, for clarity, I'd almost prefer:
```python
        if isinstance(ind, np.ndarray):
            if ind.dtype == np.bool and ind.size != self.n_cells:
                raise ValueError('Boolean array size must match the '
                                 'number of cells (%d)' % self.n_cells)
```
vs:
```python
        if (isinstance(ind, np.ndarray) and ind.dtype == np.bool and
            ind.size != self.n_cells):
            raise ValueError('Boolean array size must match the '
                             'number of cells (%d)' % self.n_cells)
```

The reason for this is `size` and `dtype` are only attributes of `np.ndarray` and changing the order of the triple condition testing will cause an error.

File: pyvista/core/pointset.py (hunk 13 above)
Comment (akaszynski) [reply pyvista/core/pointset.py]: See comment above.

File: tests/test_grid.py [hunk 14]
```diff
@@ -472,3 +472,31 @@ def test_gaussian_smooth():
     assert uniform.active_scalars_name == active
     assert uniform.active_scalars.shape == values.shape
     assert not np.all(uniform.active_scalars == values)
+
+
+@pytest.mark.parametrize('ind', [range(10), np.arange(10), np.
+                                 ones(beam.n_cells, np.bool)])
```
[the remark this answers is not in the document]
Comment (akaszynski) [reply tests/test_grid.py]: Added parameterization to the other tests.

File: pyvista/core/pointset.py (hunk 12 above)
Comment (akaszynski) [reply pyvista/core/pointset.py]: Resolved the parameterization in a5d40e51a61b470cff82ef4ef1ba4405578f4faa

**Comment by banesullivan** [comment]
FYI "blank cells" in structured grids are another useful feature here: https://github.com/pyvista/pyvista/issues/318#issuecomment-513886131

**Comment by akaszynski** [comment]
@imsodin, are we good to go?

**Comment by akaszynski** [comment]
> And just to repeat the point in a non-hidden comment (not meant to block the PR, thus approved):
> I do agree with the sentiment of filters.py being a monster of a file and the need to refactor, I do however still think that `remove_cells`/`hide_cells` in the current state of things should be filters.

I'll talk with @banesullivan about this since he implemented filters.  I feel that we shouldn't be subclassing filters, but for the time being I'll leave this alone and we can open a new PR to discuss a potential refactor of `filters.py`.  We can then move `remove_cells` and `hide_cells` to a more ideal location.

Thanks for your very thorough review of the PR.  It's good to have another critical eye here.

File: pyvista/core/pointset.py (hunk 2 above)
Comment (banesullivan) [reply pyvista/core/pointset.py]: I'm down to refactor all of that! I created that mess and I could easily reimplement the filters into standalone methods so users could call them directly from a filters submodule (e.g. `from pyvista import filters; filters.contour()`) as well as wrapped methods on classes. This would actually clean up some redefinitions and messiness with how the `MultiBlock` filters are implemented (also, this would be in line with some refactoring/migrating of PVGeo that I want to do... more on that later).

I am very hesitant to make those filters available from the top level of PyVista as that might bloat the API when new users are doing tab-complete. I'd prefer to have a new `filters` module next to the `core` module and force users to import that module much like the `examples` module. That would keep the top level only showing the most relevant API features for creating meshes, plotters, and the few other useful utilities we have available at that level.

```py
import pyvista as pv
from pyvista import filters

mesh = pv.Sphere()

deci = filters.decimate(mesh)
```

This would also make it a bit cleaner for those folks that only want to use PyVista's filters intermixed in their VTK workflows.

Comment (banesullivan) [reply pyvista/core/pointset.py]: This refactoring might create a lot of duplicated docs though as the wrapped methods will show in the autodoc of the class and then we probably want to autodoc the whole `filters` module in case folks want to use the standalone methods.

**Comment by banesullivan** [comment]
Sounds good to me - I put my thoughts in the review comments. I can handle that refactoring in a separate PR
